pandas/core/style.py

```diff
@@ -0,0 +1,777 @@
+                            uuid=uuid)
+        if trim:
+            # filter out empty styles, every cell will have a class
+            # but the list of props may just be [['', '']].
+            # so we have the neested anys below
+            trimmed = [x for x in d['cellstyle'] if
+                       any(any(y) for y in x['props'])]
+            d['cellstyle'] = trimmed
+        return self.template.render(**d)
+
+    def _update_ctx(self, attrs):
+        """
+        update the state of the Styler. Collects a mapping
+        of {index_label: ['<property>: <value>']}
+
+        attrs: Series or DataFrame
+        should contain strings of '<property>: <value>;<prop2>: <val2>'
+        Whitespace shouldn't matter and the final trailing ';' shouldn't
+        matter.
+        """
+        for row_label, v in attrs.iterrows():
+            for col_label, col in v.iteritems():
+                i = self.index.get_indexer([row_label])[0]
+                j = self.columns.get_indexer([col_label])[0]
+                for pair in col.rstrip(";").split(";"):
+                    self.ctx[(i, j)].append(pair)
+
+    def _copy(self, deepcopy=False):
+        styler = Styler(self.data, precision=self.precision,
+                        caption=self.caption, uuid=self.uuid,
+                        table_styles=self.table_styles)
+        if deepcopy:
+            styler.ctx = copy.deepcopy(self.ctx)
+        else:
+            styler.ctx = self.ctx
+        return styler
+
+    def __copy__(self):
+        """
+        Deep copy by default.
+        """
+        return self._copy(deepcopy=False)
+
+    def __deepcopy__(self, memo):
+        return self._copy(deepcopy=True)
+
+    def clear(self):
+        self.ctx.clear()
+        self._todo = []
+
+    def _compute(self):
+        '''
+        Execute the style functions built up in `self._todo`.
+
+        Relies on the conventions that all style functions go through
+        .apply or .applymap. The append styles to apply as tuples of
+
+        (application method, *args, **kwargs)
+        '''
+        r = self
+        for func, args, kwargs in self._todo:
+            r = func(self)(*args, **kwargs)
+        return r
+
+    def _apply(self, func, axis=0, subset=None, **kwargs):
+        subset = slice(None) if subset is None else subset
+        subset = _non_reducing_slice(subset)
+        if axis is not None:
+            result = self.data.loc[subset].apply(func, axis=axis, **kwargs)
+        else:
+            # like tee
+            result = func(self.data.loc[subset], **kwargs)
+        self._update_ctx(result)
+        return self
+
+    def apply(self, func, axis=0, subset=None, **kwargs):
+        """
+        Apply a function, updating the HTML representation with the result.
+
+        .. versionadded:: 0.17.1
+
+        Parameters
+        ----------
+        func: function
+        axis: int, str or None
+            apply to each column (``axis=0`` or ``'index'``)
```

Comment:
should we default ``axis=None``? e.g. make the user explict about column formatting?